nvim-types/src/non_owning.rs

```diff
@@ -0,0 +1,21 @@
+use std::{marker::PhantomData, mem::ManuallyDrop};
+
+/// A non-owning value for lifetime 'a.
+///
+/// It doesn't drop the value.
+/// Used for ffi function that accept data by value, but don't destroy or move out of it.
+/// This is garaunteed to have same layout as T
+#[repr(transparent)]
+pub struct NonOwning<'a, T> {
+    // never drop the T
+    // field not being used here, is in use for repr(transparent)
+    #[allow(unused)]
+    inner: ManuallyDrop<T>,
+    lf: PhantomData<&'a ()>,
+}
+
+impl<'a, T> NonOwning<'a, T> {
+    pub unsafe fn new(value: T) -> Self {
```

Comment:
why does this have to be `unsafe`?

Reply:
it has a contract described in the type doc comment. in particular, the owning value must not be droped before `'a` . No mutable borrows should exist to value during `'a`.

Reply:
Both are already enforced by the borrow checker since `NonOwning` is dependent on the lifetime `&self`. For example this just doesn't compile
```rust
let foo = NvimString::from("foo");
let bar = foo.non_owning();
let baz = &mut foo;
```